### PR TITLE
Add functional tests for SCP operations

### PR DESCRIPTION
**Pull request by brienw**
As I was adding tests to the context work, I realized that we need base functional tests covering the SSHKit.SCP operations first, so I've added those here.

As part of this change, I have added a shared volume to the docker image as it runs. This gives us a way to compare the state of the "remote" file system inside our tests.

---

* [X] (N/A) Documented the change if necessary
* [X] Tested this PRs change (with unit and/or functional tests)
* [X] (N/A) Added this PRs change to CHANGELOG.md (in the `#master` section) if  necessary.

### CODE REVIEW

**Comment by pmeinhardt** [comment]
Very cool, thanks for adding tests for SCP @brienw! 👍 Looks good.

I had another idea on how to test uploads which I'll just throw in here to hear your thoughts (also haven't fully thought this through yet). How about checking:

```shell
$ find PATH -exec shasum -a 1 '{}' +
da39a3ee5e6b4b0d3255bfef95601890afd80709  test
da39a3ee5e6b4b0d3255bfef95601890afd80709  test/sshkit
ee98a8de5b2feb2095b3032ecc4fbc962271cf2c  test/sshkit/context_test.exs
…
```

yields the expected output?

Pros:

- works for single files and *recursive uploads* as well
- if the comparison of the output fails, ExUnit's diff will instantly show which files
  - are missing
  - have a checksum mismatch…
- no need to mount a volume into containers
- …?

Cons:

- …?

Questions:

- How to compare timestamps if the SCP `:preserve` option is used?
- …

**Comment by brienw** [comment]
I really like that. I *knew* there had to be a better way, and I kind of ignored testing the recursive option because I knew my method was non optimal. I'll rework these today.

Thanks for the feedback and ideas!

**Comment by brienw** [comment]
Ok, that took longer than expected, plus we had a holiday here in the states so I headed off to adventure in Mexico for four days. Here's some tests more aligned with your suggestion. I had to wiggle it a little bit to get something that was portable, at least between MacOS (BSD) and Linux.

Notes:

1) I had to install `perl-utils` into the docker image so that we have access to `shasum`
2) I added a `fixtures` directory to the docker image so we can have some remote test files available for downloading

Even though the tests are passing on Travis, DO NOT MERGE YET! Because...

1) The `mtime` assertion for the `preserve: true` tests for `SCP.download` actually fail for me when I run the tests on my mac. I think there is some oddity happening with timezones even though I am asking for epoch seconds when I compare. I have debugged this and we are definitely setting `atime` and `mtime` to the correct value inside `scp/download.ex`. I tried fiddling with this to no end and I started getting tunnel vision, so I'm hoping somebody else's fresh eyes may have better luck. There are so many differences between the way both `find` and `stat` work between MacOS (BSD) and Linux (GNU) that options seem limited

2) The `atime` assertions are commented out because there seems to be a situation where although we are setting `atime` to the correct value (I have debugged this a little to verify) the compared values are always off by just a few seconds between the remote file(s) and the local file(s). I'm not sure if there is a reliable way to test this assertion, but maybe you have some ideas.

I look forward to any feedback and ideas you may have

**Comment by brienw** [comment]
@pmeinhardt: bump in case you haven't seen this. Any feedback or ideas?

**Comment by pmeinhardt** [comment]
Hey @brienw, nice hearing back from you. Really hope you enjoyed your adventure in Mexico! I'll take a look now and add my 2 cents.

File: test/sshkit/scp_functional_test.exs [hunk 1]
```diff
@@ -5,33 +5,167 @@ defmodule SSHKit.SCPFunctionalTest do
...
+    assert local_output == remote_output
+
+  end
+
+  defp get_local_stat_cmd do
+    case :os.type() do
+      {:unix, :darwin} -> "stat -f '%z\t%a\t%m'"
+      _ -> "stat -c '%s\t%X\t%Y'"
+    end
+  end
```
Comment (pmeinhardt) [inline test/sshkit/scp_functional_test.exs]: As for the issue of getting different timestamps on OS X and Alpine:

I've run a quick sanity check by mounting the project directory as a volume:

```shell
docker run --rm -it --volume $(pwd):/src sshkit-test-sshd /bin/sh
```

Then got the timestamps on both OS X and the shell in the container:

```shell
# OS X
❯ stat -f '%a %m' README.md 
1495494246 1495494246

# Alpine container
❯ stat -c '%X %Y' /src/README.md
1495494246 1495494246
```

Since they're exactly the same, it makes me think the setting of timestamps in our SCP up-/download implementation may not have the exact same result on two different machines.

Possible origin: [`:file.change_time/3`](http://erlang.org/doc/man/file.html#change_time-3) used in [`SCP.Download.touch!`](https://github.com/bitcrowd/sshkit.ex/blob/f09510f25af0c9e1a12b31a9808f2163520c452d/lib/sshkit/scp/download.ex#L194)

According to the Erlang docs, `:file.change_time` uses [`:file.write_file_info/2`](http://erlang.org/doc/man/file.html#write_file_info-2) under the hood, which uses `{time, local}` as the default for interpreting the Erlang datetime records that we're converting to (e.g. `{{2017, 5, 22}, {23, 4, 6}}`). So while [converting from epoch seconds to datetime record](https://github.com/bitcrowd/sshkit.ex/blob/master/lib/sshkit/scp/download.ex#L192-L193) and then using this in `:file.write_file_info`, the timezone may come into play.

Potential solution:

Use [`:file.write_file_info/3`](http://erlang.org/doc/man/file.html#write_file_info-3) with the `time: :posix` option instead of `:file.change_time` to set the `atime` and `mtime` of the file directly, without converting to a datetime object via `:calendar`.

File: test/support/docker/Dockerfile [hunk 2]
```diff
@@ -3,7 +3,8 @@ FROM alpine:3.5
 # Set up an Alpine Linux machine running an SSH server.
 # Autogenerate missing host keys.
 
-RUN apk add --no-cache openssh sudo
+RUN echo http://dl-cdn.alpinelinux.org/alpine/edge/main >> /etc/apk/repositories
+RUN apk add --update --no-cache openssh sudo perl-utils
```
Comment (pmeinhardt) [inline test/support/docker/Dockerfile]: – see also: https://en.wikipedia.org/wiki/Sha1sum (didn't know that `sha1sum` actually seems to be the more common case, rather than `shasum`, on Linux)

**Comment by brienw** [comment]
@pmeinhardt Here's an updated version of this branch with some cleanup and a bug fix for the `mtime` `atime` issue.

The travis build is failing because one of the updated dependencies requires elixir 1.4 but Travis is trying to run it on 1.3. I Tried updating mix.exs to use elixir 1.4 but that doesn't seem to effect what version Travis runs

File: test/sshkit/scp_functional_test.exs [hunk 3]
```diff
@@ -5,33 +5,167 @@ defmodule SSHKit.SCPFunctionalTest do
...
+    @tag boot: 1
+    test "sends a file", %{hosts: [host]} do
+      options = [port: host.port, user: host.user, password: host.password]
+      local = "test/fixtures/local_workspace/local_file.txt"
+      remote = "file.txt"
+
+      {:ok, conn} = SSH.connect(host.ip, Keyword.merge(@defaults, options))
+
+      assert :ok = SCP.upload(conn, local, remote)
+      assert verify_transfer(conn, local, remote)
```
Comment (brienw) [inline test/sshkit/scp_functional_test.exs]: `SSH.close` has been added where needed

File: test/sshkit/scp_functional_test.exs (hunk 1 above)
Comment (brienw) [inline test/sshkit/scp_functional_test.exs]: Your potential solution worked! I also moved away from using `:file` directly and instead use `File.write_stat` now

**Comment by pmeinhardt** [comment]
Nice @brienw, I will take a look in the evening. Glad the timestamps are working correctly now 🤘

File: test/sshkit/scp_functional_test.exs (hunk 3 above)
Comment (tessi) [inline test/sshkit/scp_functional_test.exs]: @pmeinhardt we should allow optionally passing lambdas to `SSH.connect`, which opens the connection only within the passed lambda:

```
SSH.connect host.ip, Keyword.merge(@defaults, options), fn(conn) ->
  SCP.upload(conn, local, remote)
end
```

what do you think?

(however, this change should not be part of this PR, it just struck me here)

Comment (tessi) [inline test/sshkit/scp_functional_test.exs]: ☝️ I've created [an issue](https://github.com/bitcrowd/sshkit.ex/issues/72) -- we can discuss there

File: test/sshkit/scp_functional_test.exs [hunk 4]
```diff
@@ -1,3 +1,178 @@
...
+    test "preserve: true", %{hosts: [host]} do
+      options = [port: host.port, user: host.user, password: host.password]
+      local = "test/fixtures/local_workspace/local_file.txt"
+      remote = "file.txt"
+
+      {:ok, conn} = SSH.connect(host.ip, Keyword.merge(@defaults, options))
+
+      assert :ok = SCP.upload(conn, local, remote, preserve: true)
+      assert verify_mode(conn, local, remote)
+      # assert verify_atime(conn, local, remote)
```
Comment (tessi) [inline test/sshkit/scp_functional_test.exs]: is this a leftover or meant to be tested?

File: test/sshkit/scp_functional_test.exs [hunk 5]
```diff
@@ -1,3 +1,178 @@
...
+      options = [port: host.port, user: host.user, password: host.password]
+      remote = "/fixtures/file.txt"
+      local = create_random_path()
+      # on_exit fn -> File.rm(local) end
+
+      {:ok, conn} = SSH.connect(host.ip, Keyword.merge(@defaults, options))
+
+      assert :ok = SCP.download(conn, remote, local, preserve: true)
+      assert verify_mode(conn, local, remote)
+      # assert verify_atime(conn, local, remote)
```
Comment (tessi) [inline test/sshkit/scp_functional_test.exs]: is this a leftover or meant to be tested? (same in line 101)

File: test/sshkit/scp_functional_test.exs (hunk 4 above)
Comment (tessi) [inline test/sshkit/scp_functional_test.exs]: > The `atime` assertions are commented out because there seems to be a situation where although we are setting `atime` to the correct value (I have debugged this a little to verify) the compared values are always off by just a few seconds between the remote file(s) and the local file(s). I'm not sure if there is a reliable way to test this assertion, but maybe you have some ideas.

i've just seen your comment above. is it still happening or did you forget to remove the comment? I will try to locally reproduce the error.

File: test/test_helper.exs [hunk 6]
```diff
@@ -12,6 +12,17 @@ unless Docker.ready? do
   exit({:shutdown, 1})
 end
 
+{output, _} = System.cmd("sha1sum", ["--version"])
```
Comment (tessi) [inline test/test_helper.exs]: i like that you try to help the user to find out (s)he didn't install `sha1sum`, but (at least for me on macos) i get the following error in that case:

```
mix test
** (ErlangError) erlang error: :enoent
    (elixir) lib/system.ex:564: System.cmd("sha1sum", ["--version"], [])
    test/test_helper.exs:15: (file)
    (elixir) lib/code.ex:370: Code.require_file/2
    (elixir) lib/enum.ex:645: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:645: Enum.each/2
```

The code seems to fail before your nice error message gets rendered.

File: test/sshkit/scp_functional_test.exs (hunk 4 above)
Comment (brienw) [inline test/sshkit/scp_functional_test.exs]: Ooops, i'll fix these, yes, these were meant to be uncommented again. i'll update this PR

**Comment by tessi** [comment]
Hey @brienw regarding the `shasum`, we did some commits in `bitcowd:brien/scp_functional_tests` to detect wether to use `shasum` or `sha1sum`.

Find them [here](https://github.com/bitcrowd/sshkit.ex/commits/brien/scp_functional_tests) -- feel free to add these commits to your PR (it felt rude to just commit to your branch :))

File: test/sshkit/scp_functional_test.exs (hunk 4 above)
Comment (brienw) [inline test/sshkit/scp_functional_test.exs]: Just took a look at this again, and yeah, the issue still exists where atime doesn't ever match. I will just remove the `atime` tests for now in the interest of moving this PR forward

**Comment by brienw** [comment]
Oh nice, i just saw your last comment @tessi, I'll get that shasum/sha1sum commit into this branch as well

**Comment by tessi** [comment]
Thanks @brienw 🎉 

I've created [a separate issue](https://github.com/bitcrowd/sshkit.ex/issues/78) for the atime-test. I'm still curious to find out what causes this. Let's see if we find time for a fix.
